src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseHeaderMap.cs

```diff
@@ -47,59 +75,152 @@ private CoseHeaderMap(bool isReadOnly)
+        /// <exception cref="InvalidOperationException">The <see cref="CoseHeaderMap"/> is read-only.</exception>
         public void Add(CoseHeaderLabel label, ReadOnlySpan<byte> value) => Add(label, CoseHeaderValue.FromBytes(value));
 
+        /// <inheritdoc/>
         public bool ContainsKey(CoseHeaderLabel key) => _headerParameters.ContainsKey(key);
 
+        /// <inheritdoc/>
         public bool TryGetValue(CoseHeaderLabel key, out CoseHeaderValue value) => _headerParameters.TryGetValue(key, out value);
 
+        /// <summary>
+        /// Removes all keys and values from the <see cref="CoseHeaderMap"/>.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The <see cref="CoseHeaderMap"/> is read-only.</exception>
         public void Clear()
         {
             ValidateIsReadOnly();
             _headerParameters.Clear();
         }
 
+        /// <inheritdoc/>
         public bool Contains(KeyValuePair<CoseHeaderLabel, CoseHeaderValue> item)
             => HeaderParametersAsCollection.Contains(item);
 
+        /// <inheritdoc/>
         public void CopyTo(KeyValuePair<CoseHeaderLabel, CoseHeaderValue>[] array, int arrayIndex)
             => HeaderParametersAsCollection.CopyTo(array, arrayIndex);
 
+        /// <inheritdoc/>
         public IEnumerator<KeyValuePair<CoseHeaderLabel, CoseHeaderValue>> GetEnumerator()
             => _headerParameters.GetEnumerator();
 
+        /// <inheritdoc/>
         IEnumerator IEnumerable.GetEnumerator()
             => _headerParameters.GetEnumerator();
 
+        /// <summary>
+        /// Removes the value with the specified label from the <see cref="CoseHeaderMap"/>.
+        /// </summary>
+        /// <param name="label">The label of the element to remove.</param>
+        /// <returns><see langword="true"/> if the element is successfully found and removed; otherwise, <see langword="false"/>. This method returns false if <paramref name="label"/> is not found in the map.</returns>
+        /// <exception cref="InvalidOperationException">The <see cref="CoseHeaderMap"/> is read-only.</exception>
         public bool Remove(CoseHeaderLabel label)
         {
             ValidateIsReadOnly();
             return _headerParameters.Remove(label);
         }
 
+        /// <summary>
+        /// Removes the first occurrence of a specific object from the <see cref="CoseHeaderMap"/>.
+        /// </summary>
+        /// <param name="item">The object to remove from the map.</param>
+        /// <returns><see langword="true"/> if item was successfully removed from the map; otherwise, <see langword="false"/>. This method also returns <see langword="false"/> if item is not found in the map.</returns>
+        /// <exception cref="InvalidOperationException">The <see cref="CoseHeaderMap"/> is read-only.</exception>
         public bool Remove(KeyValuePair<CoseHeaderLabel, CoseHeaderValue> item)
```

Comment:
Does this match only on the key, or does it require the value to also match?  That's probably worth mentioning (and having a test for, if we don't).

Reply:
I think both key and value need to match. See: https://github.com/dotnet/runtime/blob/9865cc7a89a4a613a47986b8268a10bc37adbd65/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L227-L237

Reply:
I wonder what the default comparer for ReadOnlyMemory is.  Probably an exact-ref match, vs a content match.  Sounds like we need to have three test cases for this:

* Remove using an item that the enumerator produced (should remove it)
* Remove using a manual KeyValuePair that has a key that's present, and a value that is wrong.
* Remove using a manual KeyValuePair that has a key that's present, and a value that's a copy of the existing value.

If we're not happy with the truth table of the outcome, then maybe we should also have this be an explicit implementation and not a generally public method.

Reply:
> I wonder what the default comparer for ReadOnlyMemory is. 

Why is that important? We don't use ROMemory directly on the CoseHeaderMap anymore. `TValue` in `EqualityComparer<TValue>.Default` is `CoseHeaderValue`, and its Equals impl. is `EncodedValue.Span.SequenceEqual(other.EncodedValue.Span)`.

> If we're not happy with the truth table of the outcome

|     Case                                                               |     Result    |
|------------------------------------------------------------------------|---------------|
|     Remove using one   enumerator item                                 |     True      |
|     Remove using new   KVP with key matching and value not matching    |     False     |
|     Remove using new   KVP with key not matching and value matching    |     False     |
|     Remove using new   KVP with both key and value matching            |     True      |

> Remove using a manual KeyValuePair that has a key that's present, and a value that's a copy of the existing value.

You mean that the value is a copy of an existing value of a different KVP? I don't think this is a good scenario, the items composing the kvp will always be copies since both are readonly structs.

Reply:
Ah, fair.  Still worth having the docs say that it removes if the label is present and the value matches.

Reply:
I like what [Dictionary<TKey,TValue>.ICollection<KeyValuePair<TKey,TValue>>.Remove Method](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.system-collections-generic-icollection-system-collections-generic-keyvaluepair-tkey-tvalue---remove?view=net-6.0) does for clarifying this in its `<returns>`:
> true if the key and value represented by keyValuePair is successfully found and removed; otherwise, false.